datanode/candlesv2/candle_updates.go

```diff
@@ -33,99 +34,83 @@ type subscribeRequest struct {
 ) {
+	observerName := fmt.Sprintf("candles(%s)", candleID)
 	ces := &CandleUpdates{
-		log:             log,
-		candleSource:    candleSource,
-		candleID:        candleID,
-		config:          config,
-		subscribeChan:   make(chan subscribeRequest),
-		unsubscribeChan: make(chan string),
+		log:          log,
+		candleSource: candleSource,
+		candleID:     candleID,
+		config:       config,
+		observer:     utils.NewObserver[entities.Candle](observerName, log, 10, 10),
 	}
 
+	ces.running.Store(true)
 	go ces.run(ctx)
 
 	return ces, nil
 }
 
+func (s *CandleUpdates) IsRunning() bool {
+	return s.running.Load()
+}
+
 func (s *CandleUpdates) run(ctx context.Context) {
-	subscriptions := map[string]chan entities.Candle{}
-	defer closeAllSubscriptions(subscriptions)
+	defer s.running.Store(false)
+	defer s.observer.UnsubscribeAll()
 
 	ticker := time.NewTicker(s.config.CandleUpdatesStreamInterval.Duration)
-	var lastCandle *entities.Candle
 
 	for {
 		select {
 		case <-ctx.Done():
 			return
 		case <-ticker.C:
-			if len(subscriptions) > 0 {
-				candles, err := s.getCandleUpdates(ctx, lastCandle)
+			if s.observer.GetSubscribersCount() > 0 {
+				candles, err := s.getCandleUpdates(ctx, s.lastCandle.Load())
 				if err != nil {
 					s.log.Errorf("failed to get candles, closing stream for candle id %s: %w", s.candleID, err)
 					return
 				}
 
 				if len(candles) > 0 {
-					lastCandle = &candles[len(candles)-1]
+					s.lastCandle.Store(&candles[len(candles)-1])
 				}
 
-				s.sendCandles(candles, subscriptions)
+				s.observer.Notify(candles)
 			} else {
-				lastCandle = nil
-			}
-		case subscription := <-s.subscribeChan:
-			subscriptions[subscription.id] = subscription.out
-			if lastCandle != nil {
-				s.sendCandles([]entities.Candle{*lastCandle}, map[string]chan entities.Candle{subscription.id: subscription.out})
+				s.lastCandle.Store(nil)
 			}
-		case id := <-s.unsubscribeChan:
-			removeSubscription(subscriptions, id)
 		}
 	}
 }
 
-func removeSubscription(subscriptions map[string]chan entities.Candle, subscriptionID string) {
-	if _, ok := subscriptions[subscriptionID]; ok {
-		close(subscriptions[subscriptionID])
-		delete(subscriptions, subscriptionID)
-	}
-}
-
-func closeAllSubscriptions(subscribers map[string]chan entities.Candle) {
-	for _, subscriber := range subscribers {
-		close(subscriber)
-	}
-}
-
-// Subscribe returns a unique subscription id and channel on which updates will be sent.
-func (s *CandleUpdates) Subscribe() (string, <-chan entities.Candle) {
-	out := make(chan entities.Candle, s.config.CandleUpdatesStreamBufferSize)
+func (s *CandleUpdates) Subscribe() (<-chan []entities.Candle, uint64) {
+	ch, id := s.observer.Subscribe()
 
-	nextID := atomic.AddUint64(&s.nextSubscriptionID, 1)
-	subscriptionID := fmt.Sprintf("%s-%d", s.candleID, nextID)
-	s.subscribeChan <- subscribeRequest{
-		id:  subscriptionID,
-		out: out,
+	if lastCandle := s.lastCandle.Load(); lastCandle != nil {
```

Comment:
this is I think is a race condition that could result in the same candle being sent twice, or worse, see 2 scenarios below, this is why the last candle logic was in the same state space as the per candle go routine in the current implementation: - 

Subscribe thread -> add observer
Send Thread-> store last candle 
Subscribe thread -> reads last candle
Send Thread->  Sends a shed ton of new candles, including to the new observer
Subscribe thread -> Sends now very old last candle to the new observer

Or another less drastic but similar possibility:

Subscribe thread -> add observer
Send Thread-> store last candle 
Subscribe thread -> send last candle
Send Thread -> sends same last candle to subscriber

Reply:
You're quite right. I've fixed this by adding a `SubscribeAndNotify` method to the Notifier (previously called Observer) and using that.